maintainers: add github user IDs in service to NixOS/rfcs#39

grahamc opened the pull request:
##### Motivation for this change

Begin implementing NixOS/rfcs#39 : unprivileged maintainer team.

Step one: Track user's GitHub IDs alongside their github username (re: https://github.com/NixOS/rfcs/blob/master/rfcs/0039-unprivileged-maintainer-teams.md#changes-to-maintainersmaintainer-listnix)

This list is generated by taking the usernames in the list and looking up their ID, via https://github.com/grahamc/rfc39/blob/master/src/main.rs#L79

There is a question of have any of these usernames changed hands? Hard to verify. I can tell you these errors do occur:

```
Aug 09 00:05:51.420 INFO Getting ID for user, github_account: hamhut1066, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:51.633 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: hamhut1066, exec-mode: BackfillIDs, module: rfc39:112
Aug 09 00:05:51.634 INFO Getting ID for user, github_account: siddharthist, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:51.693 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: siddharthist, exec-mode: BackfillIDs, module: rfc39:112
Aug 09 00:05:51.694 INFO Getting ID for user, github_account: tobimpub, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:51.768 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: tobimpub, exec-mode: BackfillIDs, module: rfc39:112
Aug 09 00:05:51.769 INFO Getting ID for user, github_account: f-breidenstein, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:51.840 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: f-breidenstein, exec-mode: BackfillIDs, module: rfc39:112
Aug 09 00:05:51.841 INFO Getting ID for user, github_account: ma9e, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:51.909 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: ma9e, exec-mode: BackfillIDs, module: rfc39:112
Aug 09 00:05:51.910 INFO Getting ID for user, github_account: cf6b88f, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:51.980 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: cf6b88f, exec-mode: BackfillIDs, module: rfc39:112
Aug 09 00:05:51.981 INFO Getting ID for user, github_account: nfjinjing, exec-mode: BackfillIDs, module: rfc39:100
Aug 09 00:05:52.052 WARN Error fetching ID for user, e: 404 Not Found: 'Not Found', github_account: nfjinjing, exec-mode: BackfillIDs, module: rfc39:112
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

Infinisil commented:
Aw man, really disappointed that github [allows reuse of usernames](https://help.github.com/en/articles/changing-your-github-username#about-username-changes). This can make user links in commit messages very misleading. This PR wouldn't be needed if GitHub didn't allow that.

edolstra commented on maintainers/maintainer-list.nix:
```diff
@@ -7,6 +7,7 @@
 
       # Optional
       github = "GithubUsername";
+      github_id = your-github-id;
```
Should be `githubId`.

grahamc replied:
Updated!

grahamc committed:
maintainer: siddharthist: new github is langston-barrett:

https://github.com/NixOS/nixpkgs/commit/c40a0b50f759be0160eb7e16931f6de394c4939b

grahamc committed:
maintainer: hamhut1066: new github is moredhel

https://github.com/NixOS/nixpkgs/commit/e0b81b9c3cebf46f6b3f7da1760fd60597f12ff0

grahamc committed:
maintainer: f-breidenstein: new github is fleaz

https://github.com/NixOS/nixpkgs/commit/6a2ead51ad9167e1a42e9d508920dc148f6d2cec

grahamc committed:
maintainer: ma9e: new github is furrycatherder

https://github.com/NixOS/nixpkgs/commit/8ba7aa0ed102e2d2c26ea1fcbd46bf811722be69

grahamc committed:
maintainer: tobim: new github is tobim

https://github.com/NixOS/nixpkgs/commit/4930caae80618ff9f7e752d40efb4acc809eb463

grahamc committed:
maintainer: cf6b88f: delete github record, their account is gone

grahamc committed:
maintainer: nfjinjing: delete github record, their account is gone

grahamc commented:
I updated the recorded GitHub account for @langston-barrett, @moredhel, @fleaz, @furrycatherder, and @tobim by finding the commit which added them to the maintainer list, and using the new GitHub username associated with their account.

This leaves two maintainers I couldn't track down, `cf6b88f` and `nfjinjing`. I've deleted their `github` record since they no longer have accounts.

FRidh commented:
I think we should go ahead with this. There is indeed the risk that the username (and thereby the ID that @grahamc found) corresponds now to a different user, however the impact such users can have is limited to the GitHub "Read" permission level or possibly later "Triage" level. In my opinion such a risk is acceptable.

furrycatherder commented:
The change is correct in my case, at least. Thanks for taking care of this 🙂

tobim commented:
It is correct in my case as well, sorry for not taking care of that.

grahamc commented:
Thanks @FRidh, I agree. I'm going to do a bit more looking in to the IDs I can't be 100% sure of, but let's plan on merging this (or a regeneration of this PR) today.

grahamc committed:
maintainers: add github IDs to the template, in service of NixOS/rfcs#39

grahamc commented:
The original commit adding all the IDs was +1,118. I've deleted that commit, replacing it with one adding the github ID parts to the template only. I started to regenerate the list with only IDs I'm totally confident on, but hit the rate limit which I think means the universe is telling me to do at least 37 minutes of housework.

FRidh commented:
Use the graphql api, generally speaking you get more out of it  :)

grahamc committed:
maintainers: add github IDs to most users

For each contributor's github handle, get that account name's GitHub ID.

Then, «git blame» the maintainer file and ensure GitHub agrees the commit which added the maintainer to the list was authored by the GitHub ID we've recorded.

grahamc commented:
The new commit adding GitHub IDs has 1,019 additions. I'm inclined to leave the remaining ~100 un-updated for now.

grahamc commented:
Note I've verified these names by:

1. evaluating the maintainer.nix, and getting the position of each username in the list: https://github.com/grahamc/rfc39/commit/758426b2f814932ec50c7e0946b72fabe5597002#diff-50f329b544176dd430de203797d46d2cR345
2. running git-blame on the maintainer.nix https://github.com/grahamc/rfc39/commit/758426b2f814932ec50c7e0946b72fabe5597002#diff-50f329b544176dd430de203797d46d2cR281
3. finding the commit which modified the line the maintainer is on https://github.com/grahamc/rfc39/commit/758426b2f814932ec50c7e0946b72fabe5597002#diff-50f329b544176dd430de203797d46d2cR137 (plus support for looking "behind" alphabetization and reformats)
4.  fetching the commit from github, where they'll give us the author's github ID (they track it) https://github.com/grahamc/rfc39/commit/758426b2f814932ec50c7e0946b72fabe5597002#diff-50f329b544176dd430de203797d46d2cR205
5. seeing if the author matches the github ID we have: https://github.com/grahamc/rfc39/commit/758426b2f814932ec50c7e0946b72fabe5597002#diff-50f329b544176dd430de203797d46d2cR207-R221
6. only if these match exactly, do we trust their ID is correct and have it added to the maintainer list https://github.com/grahamc/rfc39/commit/758426b2f814932ec50c7e0946b72fabe5597002#diff-40582676543fe253534e493c89a2adf4R68
